stock_picking_package_preparation_line/models/stock_picking_package_preparation_line.py

```diff
@@ -0,0 +1,243 @@
+            name = self.product_id.name_get()
+            if name:
+                self.name = name[0][1]
+            self.product_uom = self.product_id.uom_id.id
+
+    def _prepare_lines_from_pickings(self, picking_ids):
+        lines = []
+        if not picking_ids:
+            return lines
+        picking_model = self.env['stock.picking']
+        for picking in picking_model.browse(picking_ids):
+            for move_line in picking.move_lines:
+                # ----- If stock move is cancel, don't create package
+                #       preparation line
+                if move_line.state == 'cancel':
+                    continue
+                # ----- search if the move is related with a
+                #       PackagePreparationLine, yet. If not, create a new line
+                if not self.search([('move_id', '=', move_line.id)],
+                                   count=True):
+                    lines.append({
+                        'move_id': move_line.id,
+                        'name': move_line.name,
+                        'product_id': move_line.product_id.id,
+                        'product_uom_qty': move_line.product_uom_qty,
+                        'product_uom': move_line.product_uom.id,
+                        })
+        return lines
+
+    @api.multi
+    def get_move_data(self):
+        self.ensure_one()
+        return {
+            'name': self.name,
+            'product_id': self.product_id.id,
+            'product_uom_qty': self.product_uom_qty,
+            'product_uom': self.product_uom.id,
+            }
+
+
+class StockPickingPackagePreparation(models.Model):
+
+    _inherit = 'stock.picking.package.preparation'
+
+    FIELDS_STATES = {'done': [('readonly', True)],
+                     'in_pack': [('readonly', True)],
+                     'cancel': [('readonly', True)]}
+
+    picking_type_id = fields.Many2one(related='picking_ids.picking_type_id')
+    line_ids = fields.One2many(
+        'stock.picking.package.preparation.line',
+        'package_preparation_id',
+        string='Details',
+        copy=False,
+        states=FIELDS_STATES)
+
+    @api.multi
+    def _update_line_ids(self):
+        # Update package lines according to pickings.
+        # Lines with move_id will be updated
+        # Lines without move_id not be touched
+        line_model = self.env['stock.picking.package.preparation.line']
+        for pack in self:
+            for line in pack.line_ids:
+                if line.move_id:
+                    line.unlink()
+            package_preparation_lines = (
+                line_model._prepare_lines_from_pickings(
+                    [p.id for p in pack.picking_ids]))
+            if package_preparation_lines:
+                for line_vals in package_preparation_lines:
+                    line_vals['package_preparation_id'] = pack.id
+                    line_model.create(line_vals)
+
+    @api.model
+    def create(self, values):
```

Comment:
needs fixing: missing decorator  `@api.returns('self', lambda value: value.id)` (otherwise you're breaking the chain call from v7 api)